{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulations with selection for SI"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble\n",
    "\n",
    "Here we perform simulations with a selection coefficient $s=0.1$ instead of using the \n",
    "estimated $s$-values from the fit.\n",
    "\n",
    "Similar to selection.ipynb, this notebook is used to obtain \n",
    "1. create input files for simulation(parallel simulation is done on a computer cluster), see [Simulation](#methods)</a>, and\n",
    "2. analyze the simulation data (get data from simulation output files and create plots), see  [Results](#results)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='init'>Import libraries, global settings, and load experimental data</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import module,plot,data\n",
    "import pickle\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "import re\n",
    "import pandas as pd\n",
    "from functools import partial\n",
    "from matplotlib.lines import Line2D\n",
    "import string\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.set_printoptions(precision=3)\n",
    "font = {'family' : 'Arial','size'   : 8}\n",
    "matplotlib.rc('font', **font)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load experimental data from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = data.loaddata(datafile=\"expdata/plasmid-exp.xlsx\")\n",
    "dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_chromosome = data.loaddata(datafile=\"expdata/chrom-exp.xlsx\",header=1)\n",
    "dataset_chromosome.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load carrying capacities from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Nc=pd.read_csv('expdata/Nc_plasmid.csv', index_col=0).squeeze()\n",
    "Nc_chromosome=pd.read_csv('expdata/Nc_chromosome.csv', index_col=0).squeeze()\n",
    "print('Carrying capacities loaded')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='methods'>Simulation</a>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialise simulation parameters and write to input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixedparameters={\n",
    "    'n' : 15,\n",
    "    'rep' : 'reg',\n",
    "    'Nc' : Nc[1:].values.tolist(), # use day 1-30 for simulations\n",
    "    'D' : 30 # simulate for 30 days, starting at day 1\n",
    "    }\n",
    "\n",
    "fixedparameters_chromosome={\n",
    "    'n' : 1,\n",
    "    'rep' : 'reg',\n",
    "    'Nc' : Nc_chromosome[1:].values.tolist(), # use day 1-30 for simulations\n",
    "    'D' : 30 # simulate for 30 days, starting at day 1\n",
    "    }\n",
    "\n",
    "s=0.1\n",
    "s_chromosome=.1\n",
    "\n",
    "repindex_list=list(range(1,12+1))\n",
    "f_list=[ # chosen by comparison with initial frequencies at day 1\n",
    "    10**(-4),  10**(-4),  10**(-4),  10**(-4) , \n",
    "    10**(-6),  10**(-6),  10**(-6),  10**(-6),  \n",
    "    10**(-7),  10**(-7),  10**(-7),  10**(-7)  \n",
    "]\n",
    "def rwt_list(s):\n",
    "    return [  \n",
    "    1,      1-s,        1,          1-s,\n",
    "    1,      1-s,        1,          1-s,\n",
    "    1,      1-s,        1,          1-s\n",
    "    ]\n",
    "b_list=[    \n",
    "    0.01,   0.01,       0.001,      0.001,\n",
    "    0.01,   0.01,       0.001,      0.001,\n",
    "    0.01,   0.01,       0.001,      0.001,\n",
    "    ]\n",
    "\n",
    "seeds=range(100)\n",
    "\n",
    "def parset(system,s,par):\n",
    "    repindex,seed=par\n",
    "    f=f_list[repindex-1]\n",
    "    rwt=rwt_list(s)[repindex-1]\n",
    "    b=b_list[repindex-1]\n",
    "    return {\"repindex\":repindex,\"f\":f,\"rwt\":rwt,\"b\":b,\"seed\":seed,\n",
    "        \"file\":\"selection-SI_output/\"+re.sub('[^a-zA-Z0-9_]','_',\n",
    "                                '_'.join(map(str,\n",
    "                                (\"system\",system,\"repindex\",repindex,\"seed\",seed)))\n",
    "        )}\n",
    "\n",
    "l=list(itertools.product(repindex_list,seeds))\n",
    "parsets=list(map(partial(parset,'plasmid',s),l))\n",
    "\n",
    "l_chromosome=list(itertools.product(repindex_list,seeds))\n",
    "parsets_chromosome=list(map(partial(parset,'chromosome',s_chromosome),l))\n",
    "\n",
    "print('initialized ', len(parsets), ' parameter combinations for plasmid simulation')\n",
    "print('initialized ', len(parsets_chromosome), ' parameter combinations for chromosome simulation')\n",
    "\n",
    "print('repindex_list:', repindex_list)\n",
    "print('seeds:', seeds)\n",
    "\n",
    "print('Example parameter input', fixedparameters | parsets[0])\n",
    "\n",
    "# write plasmid and chromosome parametersets to input files\n",
    "j=0\n",
    "for i in range(len(parsets)):\n",
    "    with open( \"selection-SI_input/inputfile_\"+str(j), 'w') as file:\n",
    "        file.write(json.dumps(fixedparameters | parsets[i]))\n",
    "    j+=1\n",
    "for i in range(len(parsets)):\n",
    "    with open( \"selection-SI_input/inputfile_\"+str(j), 'w') as file:\n",
    "        file.write(json.dumps(fixedparameters_chromosome | parsets_chromosome[i]))\n",
    "    j+=1\n",
    "print(\"Input files saved. \")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testing simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing:\n",
    "\n",
    "print('simulation test')\n",
    "p=json.load(open( \"selection-SI_input/inputfile_1400\", 'r'))\n",
    "print(p['rwt'],p['f'],p['n'])\n",
    "del p[\"repindex\"]\n",
    "\n",
    "sim=module.stochbottleSim(**p)\n",
    "plot.plot_interdays(sim)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load and plot all simulations to check that outputfiles exist f.a. parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots()\n",
    "# i0=000\n",
    "for i,p in enumerate(parsets[:]+parsets_chromosome[:]):\n",
    "# for i,p in enumerate(parsets_chromosome[:]):\n",
    "    # print(p)\n",
    "    try:\n",
    "        sim=pickle.load(open(p['file'],'rb'))\n",
    "        # print(sim)\n",
    "        plot.plot_interdays(sim,figax=(fig,ax),\n",
    "            lim=(1e-7*0.9,1e-0/0.9),alpha=1)\n",
    "    except:\n",
    "        if not i%100 : print('Could not find '+p['file'], 'parset no: ', i) \n",
    "    \n",
    "# print parset\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='results'>Results</a>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Figure panel B (Survival of chromosome and plasmid allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Style specifications\n",
    "marker={'plasmid':'o','chromosome':'o'}\n",
    "color={'plasmid':'black','chromosome':'black'}\n",
    "linestyle={'plasmid':'solid','chromosome':'dotted'}\n",
    "dashes={'plasmid':(None,None),'chromosome':(None,None)} #(1.,1.)\n",
    "markersize=2\n",
    "linewidth=1.\n",
    "linewidthmarker=.5\n",
    "markeredgewidth=.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots(nrows=3,ncols=4,figsize=(16/2.54,10/2.54),sharex=True,sharey=True)\n",
    "plt.subplots_adjust(left=1.6/16,bottom=1.1/10,right=(1.6+3.2*4+0.4*3)/16,top=(1.1+2.5*3+0.4*2)/10,\n",
    "        wspace=0.15, hspace=0.15)\n",
    "fig.text(0.53, 0.02, 'Number of transfer', ha='center', va='center')\n",
    "fig.text(0.03, 0.52, 'Cell frequencies', ha='center', va='center', rotation='vertical')\n",
    "\n",
    "c={ 'host' : 'black',\n",
    "    'het': '#008b8bff',\n",
    "    'hom': 'violet',\n",
    "}\n",
    "l={ 'host' : 1.,\n",
    "    'het': 1,\n",
    "    'hom': 1,\n",
    "}\n",
    "ls={ 'host' :'solid',\n",
    "    'het': 'dotted',\n",
    "    'hom': 'dashed',\n",
    "}\n",
    "\n",
    "zip_repsysax={ \n",
    "            \"groupselection\" : zip(\n",
    "                [2,2,4,4,6,6,8,8,10,10,12,12,],\n",
    "                ['chromosome' if i%2 else 'plasmid'  for i in range(12)],\n",
    "                ax.flat),\n",
    "            \"groupsystem\": zip(\n",
    "                [2,4,6,8,10,12,2,4,6,8,10,12],\n",
    "                ['plasmid' if i<=6 else 'chromosome' for i in range(1,12+1)],\n",
    "                ax.flat),\n",
    "            }\n",
    "group=\"groupselection\"\n",
    "\n",
    "for rep,system,ax1 in zip_repsysax[group]:\n",
    "\n",
    "    i0=int((rep-1)*100)\n",
    "    parsetssystem={'plasmid':parsets,'chromosome':parsets_chromosome}\n",
    "    # print(i0,parsetssystem[system][i0]['file'])\n",
    "    sims=[]\n",
    "    for i in parsetssystem[system][i0:i0+100]:\n",
    "        try:\n",
    "            sims.append(pickle.load(open(i['file'],'rb')))\n",
    "        except:\n",
    "            print('no file:', i['file'])\n",
    "    genotypes=['host','het'] if system=='plasmid' else ['host']\n",
    "\n",
    "    for genotype in genotypes:\n",
    "        logfreqs=[]\n",
    "        freqs=[]\n",
    "        for i in range(len(sims)):\n",
    "            ts_eod=sims[i][1]\n",
    "            with np.errstate(divide='ignore'):\n",
    "                if genotype=='het':\n",
    "                    logfreq=np.log10(np.sum(ts_eod[:,1:-1],axis=-1)/np.sum(ts_eod,axis=-1))\n",
    "                    freq=np.sum(ts_eod[:,1:-1],axis=-1)/np.sum(ts_eod,axis=-1)\n",
    "                if genotype=='hom':\n",
    "                    logfreq=np.log10(ts_eod[:,-1]/np.sum(ts_eod,axis=-1))\n",
    "                    freq=ts_eod[:,-1]/np.sum(ts_eod,axis=-1)\n",
    "                if genotype=='host':\n",
    "                    logfreq=np.log10(np.sum(ts_eod[:,1:],axis=-1)/np.sum(ts_eod,axis=-1))\n",
    "                    freq=np.sum(ts_eod[:,1:],axis=-1)/np.sum(ts_eod,axis=-1)\n",
    "            logfreqs.append(logfreq)\n",
    "            freqs.append(freq)\n",
    "        logfreqs=np.array(logfreqs)\n",
    "        freqs=np.array(freqs)\n",
    "        ignorezeros=True\n",
    "        if ignorezeros:\n",
    "            logfreqs[np.isinf(logfreqs)]=np.nan\n",
    "            freqs[freqs==0]=np.nan\n",
    "        else:\n",
    "            logfreqs[np.isinf(logfreqs)]=-1000\n",
    "\n",
    "        q=np.nanquantile(logfreqs,q=[(1-.95)/2,(1-.68)/2,1-(1-.68)/2,1-(1-.95)/2],axis=0)\n",
    "        m=np.nanmean(freqs,axis=0)\n",
    "\n",
    "        if genotype in ['host']:\n",
    "            ax1.fill_between(range(1,len(q[0])+1),y1=q[0],y2=q[-1],color=c[genotype],alpha=.1,edgecolor=\"none\")\n",
    "            ax1.fill_between(range(1,len(q[0])+1),y1=q[1],y2=q[-2],color=c[genotype],alpha=.35,edgecolor=\"none\")\n",
    "\n",
    "        la={'host' :'Mutant',\n",
    "        'het': 'Heterozygote',\n",
    "        'hom': None}\n",
    "        ax1.plot(range(1,len(m)+1),np.log10(m),color=c[genotype],linewidth=l[genotype],linestyle=ls[genotype],label=la[genotype])\n",
    "\n",
    "    if system=='plasmid':\n",
    "        data=dataset.loc[(dataset['replicate']==rep) * (dataset['t']>=1),\n",
    "                                ['replicate.n','t','hetero_freq','homo_freq','host_freq']]\n",
    "    else:\n",
    "        data=dataset_chromosome.loc[(dataset_chromosome['replicate']==rep) * (dataset_chromosome['t']>=1),\n",
    "                                ['replicate.n','t','host_freq']]\n",
    "    data.set_index('t', inplace=True)\n",
    "    data['host_freq']=data['host_freq'].apply(np.log10)\n",
    "    data.groupby('replicate.n')['host_freq'].plot(ax=ax1, \n",
    "        color='black',\n",
    "        linestyle='none',linewidth=0.5, \n",
    "        marker=marker[system], markersize=.5,\n",
    "        alpha=.25,label='_nolegend_')\n",
    "    if 'het' in genotypes:\n",
    "        data['hetero_freq']=data['hetero_freq'].apply(np.log10)\n",
    "        data['homo_freq']=data['homo_freq'].apply(np.log10)\n",
    "        data.groupby('replicate.n')['hetero_freq'].plot(ax=ax1, \n",
    "            color='#008b8bff', \n",
    "            linestyle='None',linewidth=0.5, \n",
    "            marker=marker[system], markersize=.5,\n",
    "            alpha=.25,label='_nolegend_')\n",
    "\n",
    "    ax1.set_ylim((-8,.2));\n",
    "    ax1.set_yticks(list(range(-7,1)))\n",
    "    ax1.set_xticks([1,5,10,15,20,25,30])\n",
    "    ax1.set_xticklabels([1,5,10,15,20,25,30], fontsize=6)\n",
    "    \n",
    "    ax1.set_yticklabels([ '$10^{'+str(i)+'}$' if type(i)==int and i<0 else i for i in ['',-6,'','',-3,'','',1] ], fontsize=5)\n",
    "    # ax1.grid(linewidth=0.25)\n",
    "    ax1.yaxis.set_ticks_position('both')\n",
    "    ax1.tick_params(length=0,axis='both',which='minor')\n",
    "    ax1.tick_params(direction='in',axis='both',which='both')\n",
    "    ax1.set_xlabel('')\n",
    "\n",
    "if group==\"groupselection\":\n",
    "\n",
    "    label=['Polyploid $n_\\mathrm{rc}=15$','Monoploid $n_\\mathrm{rc}=1$','Polyploid $n_\\mathrm{rc}=15$','Monoploid $n_\\mathrm{rc}=1$']\n",
    "    for label, ax1 in zip(label,ax.flat[[0,1,2,3]]):\n",
    "        ax1.text(.55, 1.03, label, transform=ax1.transAxes, \n",
    "                size=8, weight='normal',ha='center')\n",
    "\n",
    "custom_lines = \\\n",
    "        [   \n",
    "        Line2D([0], [0], \n",
    "            color=c['host'], linestyle=ls['host'], marker='None',\n",
    "            linewidth=l['host']),\n",
    "        Line2D([0], [0], \n",
    "            color=c['het'], linestyle=ls['het'], marker='None',\n",
    "            linewidth=l['het']),\n",
    "        Line2D([0], [0], \n",
    "            color=c['host'],marker=marker['plasmid'], markersize=.5,linestyle='None',alpha=.25),\n",
    "        \n",
    "        ]\n",
    "\n",
    "numbers=['i','ii','iii','iv','v','vi','vii','viii','ix','x','xi','xii']\n",
    "numbers=['1','2','3','4','5','6','7','8','9','10','11','12']\n",
    "\n",
    "for n, ax1 in zip(numbers,\n",
    "        ax.flat):\n",
    "    ax1.text(0.00, 1.025, n, transform=ax1.transAxes, \n",
    "            size=9, weight='normal')\n",
    "for n, ax1 in zip([0,1],[ax[0,0]]):\n",
    "    ax1.text(-0.18 if n in [0,4,8] else -0.11, 1.03, string.ascii_uppercase[n], transform=ax1.transAxes, \n",
    "            size=10,weight='bold')\n",
    "\n",
    "\n",
    "ax[0,0].legend(custom_lines, ['Novel phenotype','Heterozygotes','Experiment'],fontsize=6,loc='lower right',handler_map = {tuple: matplotlib.legend_handler.HandlerTuple(None)})\n",
    "\n",
    "#ax[2,0].legend()\n",
    "\n",
    "fig.savefig('../figures-plots/plot_SI-selection_A.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots(nrows=2,ncols=4,figsize=(16/2.54,7/2.54),sharex=True,sharey=True)\n",
    "plt.subplots_adjust(left=1.6/16,bottom=1.1/7,right=(1.6+3.2*4+0.4*3)/16,top=(1.1+2.5*2+0.4*1)/7,wspace=0.15, hspace=0.15)\n",
    "fig.text(0.53, 0.05, 'Number of transfer', ha='center', va='center')\n",
    "fig.text(0.03, 0.52, 'Survival of the novel allele', ha='center', va='center', rotation='vertical')\n",
    "\n",
    "def plot_survdyn(parsets,dataset,ax_list,system='plasmid'):\n",
    "    data=dataset.loc[(dataset['t']>=1),\n",
    "                                ['replicate','replicate.n','t','host_freq']]\n",
    "\n",
    "    zeroarray_keys=['t','replicate','replicate.n','host_freq']\n",
    "    zeroarray_values=[[t,replicate,replicate_n,0.] for replicate in range(1,12+1) \n",
    "                                                    for replicate_n in range(1,6+1) \n",
    "                                                    for t in range(1,30+1)\n",
    "                    if len(dataset.loc[(dataset['t']==t) * \n",
    "                                    (dataset['replicate']==replicate) * \n",
    "                                    (dataset['replicate.n']==replicate_n),\n",
    "                                :])==0          ]\n",
    "    zeroarray_values=np.array(zeroarray_values)\n",
    "    zerodataframe=pd.DataFrame(data=zeroarray_values, columns=zeroarray_keys)\n",
    "    data=pd.concat([data,zerodataframe],ignore_index=True).sort_values(by=['replicate', 'replicate.n', 't'])\n",
    "\n",
    "    for rep,ax1 in zip([6,8,10,12],\n",
    "                    ax_list):\n",
    "        i0=int((rep-1)*100)\n",
    "        sims=[pickle.load(open(i['file'],'rb')) for i in parsets[i0:i0+100]]\n",
    "        hostcell_freqs=[]\n",
    "        for i in range(len(sims)):\n",
    "            hostcell_freqs.append(np.sum(sims[i][1][:,1:],axis=1))\n",
    "        hostcell_freqs=np.array(hostcell_freqs)\n",
    "        hostcell_freqs_nonzero=(100-np.count_nonzero(hostcell_freqs,axis=0))/100\n",
    "\n",
    "        ax1.plot(range(1,len(hostcell_freqs_nonzero)+1), 1-hostcell_freqs_nonzero,\n",
    "            color=color[system],linewidth=linewidth,linestyle=linestyle[system],alpha=1.,\n",
    "            dashes=dashes[system]\n",
    "            )\n",
    "\n",
    "        extinct_exp=data.loc[(data['replicate']==rep),\n",
    "                                ['t','host_freq']].groupby(['t']).apply(lambda x:(6-np.count_nonzero(x['host_freq']))/6)\n",
    "        ax1.errorbar(x=extinct_exp.index,y=1-extinct_exp[:],\n",
    "                #yerr=np.sqrt(extinct_exp*(1-extinct_exp)/6),\n",
    "                capsize=0.75, elinewidth=.25, markeredgewidth=markeredgewidth,\n",
    "                linestyle='None',linewidth=linewidthmarker, color=color[system], \n",
    "                marker=marker[system], markersize=markersize,alpha=1.,fillstyle='full', \n",
    "                \n",
    "                )\n",
    "        \n",
    "        ax1.set_ylim(-0.1,1.1)\n",
    "        ax1.set_xlim(-.5,21+.5)\n",
    "        ax1.set_yticks(np.linspace(0,1.0,6))\n",
    "        ax1.set_xticks([1,5,10,15,20])\n",
    "        ax1.set_yticklabels(['%i %%' % y for i,y in enumerate(np.linspace(0,100,6))],fontsize=6)\n",
    "        ax1.set_xticklabels([1,5,10,15,20],fontsize=6)\n",
    "        #ax1.grid(linewidth=0.25)\n",
    "        ax1.tick_params(length=0,axis='both',which='both')\n",
    "\n",
    "plot_survdyn(parsets,dataset,[ax[0,0],ax[0,2],ax[1,0],ax[1,2]])\n",
    "plot_survdyn(parsets_chromosome,dataset_chromosome,[ax[0,1],ax[0,3],ax[1,1],ax[1,3]],system='chromosome')\n",
    "\n",
    "custom_lines = \\\n",
    "        [   \n",
    "        \n",
    "        Line2D([0], [0], \n",
    "            color=color['plasmid'], linestyle=linestyle['plasmid'], marker='None',\n",
    "            linewidth=linewidth,dashes=dashes['plasmid']),\n",
    "        Line2D([0], [0], \n",
    "            color=color['plasmid'], linestyle='None', marker=marker['plasmid'],\n",
    "            fillstyle='full',markersize=markersize,markeredgewidth=markeredgewidth),\n",
    "        ]\n",
    "\n",
    "ax[0,0].legend(custom_lines, ['Simulation','Experiment'],fontsize=6,handler_map = {tuple: matplotlib.legend_handler.HandlerTuple(None)})\n",
    "\n",
    "label=['Polyploid $n_\\mathrm{rc}=15$','Monoploid $n_\\mathrm{rc}=1$','Polyploid $n_\\mathrm{rc}=15$','Monoploid $n_\\mathrm{rc}=1$']\n",
    "for label, ax1 in zip(label,ax.flat[[0,1,2,3]]):\n",
    "    ax1.text(.55, 1.03, label, transform=ax1.transAxes, \n",
    "            size=8, weight='normal',ha='center')\n",
    "\n",
    "numbers=['i','ii','iii','iv','v','vi','vii','viii','ix','x','xi','xii']\n",
    "numbers=['1','2','3','4','5','6','7','8','9','10','11','12']\n",
    "\n",
    "\n",
    "for n, ax1 in zip(numbers,\n",
    "        ax.flat):\n",
    "    ax1.text(0.00, 1.025, n, transform=ax1.transAxes, \n",
    "            size=9, weight='normal')\n",
    "\n",
    "for n, ax1 in enumerate([ax[0,0]]):\n",
    "    ax1.text(-0.18 , 1.02, string.ascii_uppercase[n+1], transform=ax1.transAxes, \n",
    "            size=10, weight='bold')\n",
    "\n",
    "fig.savefig('../figures-plots/plot_SI-selection_B.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
